bee-network/bee-autopeering/src/discovery/messages.rs

```diff
@@ -0,0 +1,286 @@
+    #[error("Invalid target ip address. Cause: {0}.")]
+    InvalidTargetIpAddress(AddrParseError),
+    #[error("{0}")]
+    ProtobufDecode(#[from] DecodeError),
+    #[error("{0}")]
+    ProtobufEncode(#[from] EncodeError),
+}
+
+#[derive(Clone, Copy)]
+pub(crate) struct VerificationRequest {
+    version: u32,
+    network_id: u32,
+    timestamp: u64,
+    source_addr: SocketAddr,
+    target_addr: IpAddr,
+}
+
+impl VerificationRequest {
+    pub fn new(version: u32, network_id: u32, source_addr: SocketAddr, target_addr: IpAddr) -> Self {
+        let timestamp = crate::time::unix_now_secs();
+
+        Self {
+            version,
+            network_id,
+            timestamp,
+            source_addr,
+            target_addr,
+        }
+    }
+
+    pub fn version(&self) -> u32 {
+        self.version
+    }
+
+    pub fn network_id(&self) -> u32 {
+        self.network_id
+    }
+
+    pub fn timestamp(&self) -> u64 {
+        self.timestamp
+    }
+
+    pub fn source_addr(&self) -> SocketAddr {
+        self.source_addr
+    }
+
+    pub fn from_protobuf(bytes: &[u8]) -> Result<Self, Error> {
+        let proto::Ping {
+            version,
+            network_id,
+            timestamp,
+            src_addr,
+            src_port,
+            dst_addr,
+        } = proto::Ping::decode(bytes)?;
+
+        let ip_addr: IpAddr = src_addr.parse().map_err(Error::InvalidSourceIpAddress)?;
+        let port = src_port as u16;
+
+        let source_addr = SocketAddr::new(ip_addr, port);
+        let target_addr: IpAddr = dst_addr.parse().map_err(Error::InvalidTargetIpAddress)?;
+
+        Ok(Self {
+            version,
+            network_id,
+            timestamp: timestamp as u64,
+            source_addr,
+            target_addr,
+        })
+    }
+
+    #[allow(clippy::wrong_self_convention)]
+    pub fn to_protobuf(&self) -> Result<BytesMut, EncodeError> {
+        let ping = proto::Ping {
+            version: self.version,
+            network_id: self.network_id,
+            timestamp: self.timestamp as i64,
+            src_addr: self.source_addr.ip().to_string(),
+            src_port: self.source_addr.port() as u32,
+            dst_addr: self.target_addr.to_string(),
+        };
+
+        let mut bytes = BytesMut::with_capacity(ping.encoded_len());
+        ping.encode(&mut bytes)?;
+
+        Ok(bytes)
+    }
+}
+
+impl fmt::Debug for VerificationRequest {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("VerificationRequest")
+            .field("version", &self.version)
+            .field("network_id", &self.network_id)
+            .field("timestamp", &self.timestamp)
+            .field("source_addr", &self.source_addr)
+            .field("target_addr", &self.target_addr)
+            .finish()
+    }
+}
+
+impl Request for VerificationRequest {}
+
+#[derive(Clone)]
+pub(crate) struct VerificationResponse {
+    request_hash: Vec<u8>,
```

Comment:
Isn't this supposed to have a fixed size ?

Reply:
yup, protobuf type is falsely leaking into this wrapper type.

Changed for this and the `DiscoveryResponse` type as well.